File: Braintree/src/test/java/com/braintreepayments/api/CrashReporterUnitTest.java
package com.braintreepayments.api;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;

@RunWith(RobolectricGradleTestRunner.class)
public class CrashReporterUnitTest {

    private BraintreeFragment mBraintreeFragment;
    private Thread.UncaughtExceptionHandler mDefaultUncaughtExceptionHandler;
    private CrashReporter mCrashReporter;

    @Before
    public void setup() {
        mDefaultUncaughtExceptionHandler = mock(Thread.UncaughtExceptionHandler.class);
        Thread.setDefaultUncaughtExceptionHandler(mDefaultUncaughtExceptionHandler);
        mBraintreeFragment = mock(BraintreeFragment.class);
        mCrashReporter = CrashReporter.setup(mBraintreeFragment);
    }

    @Test
    public void setup_setsCrashReporterAsTheDefaultUncaughtExceptionHandler() {
        assertTrue(Thread.getDefaultUncaughtExceptionHandler() instanceof CrashReporter);
    }

    @Test
    public void tearDown_setsDefaultUncaughtExceptionHandlerBack() {
        mCrashReporter.tearDown();

        assertFalse(Thread.getDefaultUncaughtExceptionHandler() instanceof CrashReporter);
    }

    @Test
    public void uncaughtException_doesNotSendCrashEventIfStackTraceDidNotContainBraintreeCode() {
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("test", "test", "test", 1) });

        mCrashReporter.uncaughtException(null, exception);

        verifyZeroInteractions(mBraintreeFragment);
    }

    @Test
    public void uncaughtException_forwardsExceptionToDefaultUncaughtExceptionHandlerForNonBraintreeCrashes() {
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("test", "test", "test", 1) });

        mCrashReporter.uncaughtException(null, exception);

        verify(mDefaultUncaughtExceptionHandler).uncaughtException(null, exception);
    }

    @Test
    public void uncaughtException_sendsCrashEventIfStackTraceContainsBraintreeCode() {
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("com.braintreepayments.api.CrashReporting", "test", "test", 1) });
        mCrashReporter.uncaughtException(null, exception);

        exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("com.paypal.CrashReporting", "test", "test", 1) });
        mCrashReporter.uncaughtException(null, exception);

        verify(mBraintreeFragment, times(2)).sendAnalyticsEvent("crash");
    }

    @Test
    public void uncaughtException_forwardsExceptionToDefaultUncaughtExceptionHandlerForBraintreeCrashes() {
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("com.braintreepayments.api.CrashReporting", "test", "test", 1) });

        mCrashReporter.uncaughtException(null, exception);

        verify(mDefaultUncaughtExceptionHandler).uncaughtException(null, exception);
    }

    @Test
    public void uncaughtException_doesNotCrashIfDefaultUncaughtExceptionHandlerWasNull() {
        Thread.setDefaultUncaughtExceptionHandler(null);
        mCrashReporter = CrashReporter.setup(mBraintreeFragment);
        Exception exception = new Exception();
        exception.setStackTrace(new StackTraceElement[] { new StackTraceElement("com.braintreepayments.api.CrashReporting", "test", "test", 1) });

        mCrashReporter.uncaughtException(null, exception);
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/core/CrashReporterUnitTest.kt
package com.braintreepayments.api.core

import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert.assertSame
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.lang.ref.WeakReference

@RunWith(RobolectricTestRunner::class)
@Suppress("MaxLineLength")
class CrashReporterUnitTest {

    private lateinit var braintreeClient: BraintreeClient
    private lateinit var defaultExceptionHandler: Thread.UncaughtExceptionHandler

    @Before
    fun setup() {
        braintreeClient = mockk(relaxed = true)
        defaultExceptionHandler = mockk(relaxed = true)
    }

    @Test
    fun `when start is called, CrashReporter sets itself as the thread default exception handler`() {
        val sut = CrashReporter(braintreeClient)
        sut.start()
        assertSame(sut, Thread.getDefaultUncaughtExceptionHandler())
    }

    @Test
    fun `when braintreeClient weak reference is null, uncaughtException forwards to the default handler`() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception("error")

        val sut = CrashReporter(WeakReference(null))
        sut.start()
        sut.uncaughtException(thread, exception)
        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }

    @Test
    fun `when braintreeClient weak reference is null, uncaughtException restores the original default handler`() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val sut = CrashReporter(WeakReference(null))
        sut.start()
        sut.uncaughtException(mockk(), Exception("error"))
        assertSame(defaultExceptionHandler, Thread.getDefaultUncaughtExceptionHandler())
    }

    @Test
    fun `when stack trace has no known braintree or paypal cause, uncaughtException forwards to the default handler without reporting`() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception()
        exception.stackTrace =
            arrayOf(StackTraceElement("test",
                "test",
                "test",
                1))
        val sut = CrashReporter(braintreeClient)
        sut.start()
        sut.uncaughtException(thread, exception)

        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }

    @Test
    fun `when stack trace contains a braintree class, uncaughtException reports the crash and forwards to the default handler`() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception()
        exception.stackTrace = arrayOf(
            StackTraceElement("com.braintreepayments.api.CrashReporting",
                "test",
                "test",
                1)
        )
        val sut = CrashReporter(braintreeClient)
        sut.start()
        sut.uncaughtException(thread, exception)

        verify { braintreeClient.reportCrash() }
        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }

    @Test
    fun `when stack trace contains a paypal class, uncaughtException reports the crash and forwards to the default handler`() {
        Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
        val thread = mockk<Thread>()
        val exception = Exception()
        exception.stackTrace = arrayOf(
            StackTraceElement("com.paypal.CrashReporting",
                "test",
                "test",
                1)
        )
        val sut = CrashReporter(braintreeClient)
        sut.start()
        sut.uncaughtException(thread, exception)

        verify { braintreeClient.reportCrash() }
        verify { defaultExceptionHandler.uncaughtException(thread, exception) }
    }
}
